prow/tide/tide_test.go

```diff
@@ -5016,9 +5016,92 @@ func TestIsBatchCandidateEligible(t *testing.T) {
 				OptionalContexts: []string{optionalContextName},
 			}
 
-			if actual := c.isBatchCandidateEligible(logrus.WithField("tc", tc.name), &pr, cc); actual != tc.expected {
+			if actual := c.isRetestEligible(logrus.WithField("tc", tc.name), &pr, cc); actual != tc.expected {
 				t.Errorf("expected result %t, got %t", tc.expected, actual)
 			}
 		})
 	}
 }
+
+// TestSerialRetestingConsidersPRThatIsCurrentlyBeingSRetested verifies the following sequence of events:
+// 1. Tide creates a serial retest run for a passing PR
+// 2. The status context on the PR gets updated to pending
+// 3. Another PR gets merged and changed the baseSHA, for example because it already had up-to-date tests but was missing labels
+// 4. Tide will again trigger serial retests for the passing PR (The runs from step 1 will be deleted by Plank)
+func TestSerialRetestingConsidersPRThatIsCurrentlyBeingSRetested(t *testing.T) {
+	t.Parallel()
+	configGetter := func() *config.Config {
+		return &config.Config{
+			ProwConfig: config.ProwConfig{Tide: config.Tide{
+				Queries:       config.TideQueries{{}},
+				MaxGoroutines: 1,
+			}},
+			JobConfig: config.JobConfig{PresubmitsStatic: map[string][]config.Presubmit{
+				"/": {{AlwaysRun: true, Reporter: config.Reporter{Context: "mandatory-job"}}},
+			}},
+		}
+	}
+	ghc := &fgc{}
+	mmc := newMergeChecker(configGetter, ghc)
+	mgr := newFakeManager()
+	log := logrus.WithField("test", t.Name())
+	history, err := history.New(1, nil, "")
+	if err != nil {
+		t.Fatalf("failed to construct history: %v", err)
+	}
+	c, err := newSyncController(
+		context.Background(), log, ghc, mgr, configGetter, nil, &statusController{}, history, mmc, false,
+	)
+	if err != nil {
+		t.Fatalf("failed to construct sync controller: %v", err)
+	}
+
+	// Add a successful PR to github
+	initialPR := PullRequest{}
+	initialPR.Commits.Nodes = append(initialPR.Commits.Nodes, struct{ Commit Commit }{
+		Commit: Commit{Status: CommitStatus{Contexts: []Context{
+			{
+				Context: githubql.String("mandatory-job"),
+				State:   githubql.StatusStateSuccess,
+			},
+			{
+				Context: githubql.String(statusContext),
+				State:   githubql.StatusStatePending,
+			},
+		}}},
+	})
+	ghc.prs = map[string][]PullRequest{"": {initialPR}}
+
+	// sync, this creates a new serial retest prowjob
+	if err := c.Sync(); err != nil {
+		t.Fatalf("sync failed: %v", err)
+	}
+	// ensure there is actually the retest job
+	var pjs prowapi.ProwJobList
+	if err := c.prowJobClient.List(c.ctx, &pjs); err != nil {
+		t.Fatalf("failed to list prowjobs: %v", err)
+	}
+	if n := len(pjs.Items); n != 1 {
+		t.Errorf("expected to find exactly one prowjob, got %d from list %+v", n, pjs)
+	}
+
+	// Update the context on the PR to pending just like crier would
+	initialPR.Commits.Nodes[0].Commit.Status.Contexts[0].State = githubql.StatusStatePending
```

Comment:
assuming the first context is this job makes sense to me in this scenario, however I'm not sure that it's guaranteed if someone else makes change to this test or update code logic. My main concern is the context for `tide` itself, does this make sense to you?

Reply:
Good point, I updated it to only update the context if its name matches the Prowjobs context